### ISSUE TITLE
json_tokener_parse() in master sets errno=1 "Operation not permitted"

### ISSUE THREAD

**Issue by MarcT512** [json-c/json-c#601]
json_tokener_parse() is setting errno=1 on my system with current git master, even when successful.

It did not do this with json-c v0.12.

Unfortunately I'm not  sure how to trace it further.

**Comment by MarcT512** [comment]
I did a git bisect and found the commit which introduced this behavior: 3c3b5920f77c751471cd46b92f9d3f5aa6ff2557 add uint64 data to json-c

In which there is:
```
+int json_parse_uint64(const char *buf, uint64_t *retval)
+{
+       char *end = NULL;
+       uint64_t val;
+       errno = 1;
```

I can't say I fully understand this change, but could the original errno be preserved (or use some other variable here)...?

**Comment by ploxiln** [comment]
yeah that looks wrong, I can open a PR with a minimal fix

**Comment by hawicz** [comment]
You shouldn't be checking errno unless an error is actually returned.  It's perfectly fine for it to be set to any value, and it's entirely meaningless to depend on it unless the function you've called indicates that something failed.

**Comment by hawicz** [comment]
Fyi, I merged #602 because it improved the internal logic of json_parse_uint64(), not because it results in json_tokener_parse() returning with errno==0.  Just like with the [C standard library](https://stackoverflow.com/questions/10211248/errno-value-is-not-updated/10211365#10211365), it's only valid to check errno after a failure.

**Comment by ploxiln** [comment]
I see, thanks.
